sdks/python/apache_beam/transforms/util.py

```diff
@@ -161,3 +174,215 @@ def RemoveDuplicates(pcoll):  # pylint: disable=invalid-name
+               target_batch_overhead=.1,
+               target_batch_duration=1,
+               clock=time.time):
+    if min_batch_size > max_batch_size:
+      raise ValueError("Minimum (%s) must not be greater than maximum (%s)" % (
+          min_batch_size, max_batch_size))
+    self._min_batch_size = min_batch_size
+    self._max_batch_size = max_batch_size
+    self._target_batch_overhead = target_batch_overhead
+    self._target_batch_duration = target_batch_duration
+    self._clock = clock
+    self._data = []
+    self._ignore_next_timing = False
+    self._size_distribution = Metrics.distribution(
+        'BatchElements', 'batch_size')
+    self._time_distribution = Metrics.distribution(
+        'BatchElements', 'msec_per_batch')
+    self._remainder_msecs = 0
+
+  def ignore_next_timing(self, ignore_next_timing=True):
+    """Call to indicate the next timing should be ignored.
+
+    For example, the first emit of a ParDo operation is known to be anomolous
+    due to setup that may occur.
+    """
+    self._ignore_next_timing = ignore_next_timing
+
+  @contextlib.contextmanager
+  def record_time(self, batch_size):
+    start = self._clock()
+    yield
+    elapsed = self._clock() - start
+    elapsed_msec = 1e3 * elapsed + self._remainder_msecs
+    self._size_distribution.update(batch_size)
+    self._time_distribution.update(int(elapsed_msec))
+    self._remainder_msecs = elapsed_msec - int(elapsed_msec)
+    if self._ignore_next_timing:
+      self._ignore_next_timing = False
+    else:
+      self._data.append((batch_size, elapsed))
+      if len(self._data) >= self.MAX_DATA_POINTS:
+        self._thin_data()
+
+  def _thin_data(self):
+    sorted_data = sorted(self._data)
+    odd_one_out = [sorted_data[-1]] if len(sorted_data) % 2 == 1 else []
+    # Sort the pairs by how different they are.
+    pairs = sorted(zip(sorted_data[::2], sorted_data[1::2]),
+                   key=lambda ((x1, _1), (x2, _2)): x2 / x1)
+    # Keep the top 1/3 most different pairs, average the top 2/3 most similar.
+    threshold = 2 * len(pairs) / 3
+    self._data = (
+        list(sum(pairs[threshold:], ()))
+        + [((x1 + x2) / 2.0, (t1 + t2) / 2.0)
+           for (x1, t1), (x2, t2) in pairs[:threshold]]
+        + odd_one_out)
+
+  def next_batch_size(self):
+    if self._min_batch_size == self._max_batch_size:
+      return self._min_batch_size
+    elif len(self._data) < 1:
+      return self._min_batch_size
+    elif len(self._data) < 2:
+      # Force some variety.
+      return int(max(
+          min(self._max_batch_size,
+              self._min_batch_size * self.MAX_GROWTH_FACTOR),
+          self._min_batch_size + 1))
+
+    # Linear regression for y = a + bx, where x is batch size and y is time.
+    xs, ys = zip(*self._data)
+    n = float(len(self._data))
+    xbar = sum(xs) / n
+    ybar = sum(ys) / n
+    b = (sum([(x - xbar) * (y - ybar) for x, y in self._data])
+         / sum([(x - xbar)**2 for x in xs]))
```

Comment:
<!--thread_id:cc_144436952_t; commit:85e7756335b0cf7501f62bd283b07fa9ea468a21; resolved:1-->
<!--section:context-quote-->
> **katsiapis** wrote:
> What guarantees this will never be 0 (we need this guarantee since we are dividing by it right)?

<!--section:body-->
This is forced by the len(self._data) < 2 clause above. Added more to the comment.